athenz-identity-provider-service/src/main/java/com/yahoo/vespa/hosted/athenz/instanceproviderservice/ca/CertificateSigner.java

```diff
@@ -0,0 +1,124 @@
+
+    private final PrivateKey caPrivateKey;
+    private final X500Name issuer;
+    private final Clock clock;
+
+    public CertificateSigner(KeyProvider keyProvider,
+                             AthenzProviderServiceConfig.Zones zoneConfig,
+                             String configServerHostname) {
+        this(keyProvider.getPrivateKey(zoneConfig.secretVersion()), configServerHostname, Clock.systemUTC());
+    }
+
+    CertificateSigner(PrivateKey caPrivateKey, String configServerHostname, Clock clock) {
+        this.caPrivateKey = caPrivateKey;
+        this.issuer = new X500Name("CN=" + configServerHostname);
+        this.clock = clock;
+    }
+
+    X509Certificate generateX509Certificate(PKCS10CertificationRequest certReq, String remoteHostname) {
+        assertCertificateCommonName(certReq.getSubject(), remoteHostname);
+        assertCertificateExtensions(certReq);
+
+        Date notBefore = Date.from(clock.instant());
+        Date notAfter = Date.from(clock.instant().plus(CERTIIFICATE_DURATION));
+
+        try {
+            PublicKey publicKey = new JcaPKCS10CertificationRequest(certReq).getPublicKey();
+            X509v3CertificateBuilder caBuilder = new JcaX509v3CertificateBuilder(
+                    issuer, BigInteger.valueOf(clock.millis()), notBefore, notAfter, certReq.getSubject(), publicKey)
+
+                    // Set Basic Constraints to false
+                    .addExtension(Extension.basicConstraints, false, new BasicConstraints(false));
+
+            ContentSigner caSigner = new JcaContentSignerBuilder(SIGNER_ALGORITHM).build(caPrivateKey);
+
+            return new JcaX509CertificateConverter()
+                    .setProvider(new BouncyCastleProvider())
+                    .getCertificate(caBuilder.build(caSigner));
+        } catch (Exception ex) {
+            log.log(LogLevel.ERROR, "Failed to generate X509 Certificate", ex);
+            throw new RuntimeException("Failed to generate X509 Certificate");
+        }
+    }
+
+    static void assertCertificateCommonName(X500Name subject, String commonName) {
+        List<AttributeTypeAndValue> attributesAndValues = Arrays.stream(subject.getRDNs())
+                .flatMap(rdn -> rdn.isMultiValued() ?
+                        Stream.of(rdn.getTypesAndValues()) : Stream.of(rdn.getFirst()))
+                .filter(attr -> attr.getType() == BCStyle.CN)
+                .collect(Collectors.toList());
+
+        if (attributesAndValues.size() != 1) {
+            throw new IllegalArgumentException("Only 1 common name should be set");
+        }
+
+        String actualCommonName = DERUTF8String.getInstance(attributesAndValues.get(0).getValue()).getString();
+        if (! actualCommonName.equals(commonName)) {
+            throw new IllegalArgumentException("Expected common name to be " + commonName + ", but was " + actualCommonName);
+        }
+    }
+
+    static void assertCertificateExtensions(PKCS10CertificationRequest request) {
```

Comment:
See comment about `assertCertificateCommonName`

Reply:
Renamed to `verify*`.